tools/2025-06-radius-upgrades-flux.md

```diff
@@ -0,0 +1,436 @@
+# Radius GitOps Helm Upgrades with Flux
+
+* **Author**: Will Smith (@willdavsmith)
+* **Status**: In Progress
+
+## Overview
+
+This design enables automated Radius upgrades in GitOps environments using Flux HelmRelease resources. The Radius Kubernetes controller is extended with a new reconciler that monitors Flux HelmRelease objects and validates upgrade compatibility before allowing Flux to proceed with Helm upgrades. It uses the existing preflight check system to ensure that upgrades are safe and compatible with the current cluster state.
+
+## Terms and Definitions
+
+| Term | Definition |
+|------|------------|
+| **GitOps** | A deployment methodology where infrastructure and applications are managed declaratively through Git repositories |
+| **Flux** | A CNCF graduated GitOps tool that automatically syncs Kubernetes clusters with Git repositories |
+| **HelmRelease** | A Flux custom resource that defines how Helm charts should be deployed |
+| **Preflight Checks** | Validation tests run before upgrades to ensure compatibility and prevent problematic deployments |
+| **Reconciler** | A Kubernetes controller component that continuously monitors and acts on resource changes |
+
+## Objectives
+
+### Goals
+
+- **Enable Safe GitOps Upgrades**: Allow Radius upgrades to be managed through GitOps workflows while maintaining safety guarantees
+- **Leverage Existing Preflight System**: Reuse the established preflight check infrastructure from the CLI for consistent validation
+- **Provide Opt-in Control**: Users must explicitly enable the feature to avoid unexpected behavior
+- **Maintain GitOps Principles**: Work within GitOps patterns without requiring manual intervention
+- **Ensure Observability**: Provide clear feedback through Kubernetes events and annotations
+
+### Non-Goals
+
+- Supporting GitOps tools other than Flux (future consideration)
+- Automatic rollback of failed upgrades (handled by Flux/Helm)
+- Custom preflight checks specific to GitOps (use existing registry)
+- Managing non-Radius Helm charts
+
+## User Experience
+
+### Target Users
+
+**Platform Engineers** who:
+- Manage Radius deployments in production environments
+- Use GitOps workflows for infrastructure and application deployment
+- Need automated but safe upgrade processes
+- Require audit trails and observability for compliance
+
+### User Scenarios
+
+#### Scenario 1: Successful Automated Upgrade
+```yaml
+# Platform engineer updates Radius version in Git
+apiVersion: helm.toolkit.fluxcd.io/v2
+kind: HelmRelease
+metadata:
+  name: radius
+  namespace: flux-system
+  annotations:
+    radapp.io/upgrade-enabled: "true"  # Opt-in to safety checks
```

Comment:
Is this the only safety mechanism that would block the reconciler from running the preflight checks and the upgrade?

If so, this should be disabled by defautl and then users should enable it if they want.

Reply:
I guess this is false by default.